### ISSUE TITLE
Handle 404s from NWISweb?

### ISSUE THREAD

**Issue by jread-usgs** [USGS-R/dataRetrieval#430]
This question comes from Charlie regarding the following query:
```r
huc_dataGW <- readNWISdata(huc = "19", hasDataTypeCd = "qw", parameterCd = "66604", sitetype = "GW", service = "site", seriesCatalogOutput = FALSE)
```
as of 2018-06-29, [this query](https://waterservices.usgs.gov/nwis/site/?huc=19&hasDataTypeCd=qw&parameterCd=66604&sitetype=GW&seriesCatalogOutput=FALSE&format=rdb) is a 404 but many other hucs (e.g., "01") aren't 404s as they have data. Charlie found "The line works fine for hucs that have data. I checked in NWISWeb and there are no GW sites with this pcode in huc 19."

I'm not sure what others think, but I think it might be dangerous to do anything other than fail when we hit this error code, since it means the client couldn't reach the server. From Charlie again: "The USGS NWIS water service web page (https://waterservices.usgs.gov/rest/Site-Service.html#Service) states a 404 message is _Returned if and only if the query expresses a combination of elements where data do not exist_."

If we silently handle 404s and (for example) return no data, then bad urls will look like they no data (e.g., https://waterservices.usgs.gov/nwis/faketest/ is a 404) instead of being clearly identified as bad urls.

**Comment by wdwatkins** [comment]
Yeah, I think we've tried to stay away from compensating for this kind of thing in dataRetrieval.

**Comment by ldecicco-USGS** [comment]
I'll ping the NWIS team because I agree with Charlie that it shouldn't be returning a 404....but I also think dataRetrieval shouldn't be doing anything different (404 != no data, 404 *should* mean invalid query).

**Comment by ldecicco-USGS** [comment]
Ahhh, yeah, now I get the issue. This is the one thing "waterdata" does better than "waterservices" in my opinion. "waterdata" queries that are legit and have no data come back with a nice message "No sites/data found using the selection criteria specified" a la:

https://nwis.waterdata.usgs.gov/nwis/qwdata?search_site_no=455638089034501&search_site_no_match_type=exact&multiple_parameter_cds=00060&param_cd_operator=AND&list_of_search_criteria=search_site_no,multiple_parameter_cds&group_key=NONE&sitefile_output_format=html_table&column_name=agency_cd&column_name=site_no&column_name=station_nm&inventory_output=0&rdb_inventory_output=file&TZoutput=0&pm_cd_compare=Greater%20than&radio_parm_cds=previous_parm_cds&qw_attributes=0&format=rdb&rdb_qw_attributes=expanded&date_format=YYYY-MM-DD&rdb_compression=value&qw_sample_wide=0&begin_date=2014-10-01&end_date=2015-09-30

Also, "waterservices" using waterml1 can give a similar message for data queries:
https://waterservices.usgs.gov/nwis/dv/?site=05212700&format=waterml,1.1&ParameterCd=00060&StatCd=00001&startDT=2014-01-01&endDT=2014-01-10

But useful information is returned from 404 and 400s sometimes:
https://waterservices.usgs.gov/nwis/site/?huc=19&hasDataTypeCd=qw&parameterCd=66604&sitetype=GW&seriesCatalogOutput=FALSE&format=rdb
(404...but with a html body message: "HTTP Status 404 - No sites found matching this request, server=[caas01]")

https://waterservices.usgs.gov/nwis/site/?huc=19&hasDataTypeCd=qw&parameterCd=hi&sitetype=GW&seriesCatalogOutput=FALSE&format=rdb
(400...but with a html body message "HTTP Status 400 - ParameterCd: length must be no less than 5 characters, server=[sdas01]")

**Comment by wdwatkins** [comment]
Would it be feasible and worthwhile to try to parse that html so the user sees that embedded message?   

Or print a message for the user to try that url in their browser so they can see it themselves?  Not sure what kind of tryCatch that would need

**Comment by ldecicco-USGS** [comment]
Yup, we could. I just want to verify that there's either not a better way, or if the NWIS team would consider a more standard approach.

**Comment by ldecicco-USGS** [comment]
So....for the record, it'll be straight-forward to add the 404/400 body messages...but that will not help (on it's own) reduce the need for a `tryCatch` in this instance....unless we really want to start getting into somewhat unknown territory (to me). That is...we could check for very specific messages, and then return an empty data frame in those cases....but I think there will simply be too many actual useful messages for us to do that well.

**Comment by wdwatkins** [comment]
Taking actions based on certain messages sounds like a :can_o_worms: to me

**Comment by ldecicco-USGS** [comment]
There is currently a ticket to fix this on the NWIS service side (this issue may help speed up the completion of the ticket). For now, I'm going to not consider fixing it on the dataRetrieval side of things. I'm following the ticket now, so when there is work, I can re-consider what dataRetrieval does then.
